### PR TITLE
fix currency breaking issue with currency conversion

### PR DESCRIPTION
**Pull request by b4s36t4**
### Fixed Issues

$ https://github.com/Expensify/App/issues/43004
PROPOSAL: https://github.com/Expensify/App/issues/43004#issuecomment-2145817692

### Tests

**Maybe a technical testing is required than manual QA**

1. Open any report
2. Click (+) icon > Submit expense then submit an expense
3. Open the expense report that you created it in step 2 and Verify the app is not crashed 

* extra steps for Web and Desktop only
 
4. Open your browser console and type `Onyx.merge("report_<your_report_id>",{currency: null})` then press Enter. You can get `your_report_id` from browser link.
5. Verify the app is not crashed

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

Same as Above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/59088937/221ea26e-e202-4d0f-9992-9f05fcefe59d

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/59088937/82394ca2-3349-42ef-85fa-08268cc901c7

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/59088937/3ea3fa27-85ad-4757-8d85-38a89a170210

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/59088937/770b3030-62ae-4be6-b7a1-831d15635710

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/59088937/0c61b1da-0ddd-481e-b147-6b4f0de12e80

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/59088937/ca7c165e-7c65-4e23-92f4-f85d93e30408

</details>

### CODE REVIEW

**Comment by b4s36t4** [comment]
I'll update the recordings shortly.

**Comment by b4s36t4** [comment]
@ahmedGaber93 can you take latest pull from main and execute the Onyx.merge command on a IOU report please?

I'm getting the following error, wanted to know if it's because of my commits or it's on latest itself.

> I tried on `main` but wanted to make sure again, so asking you.

Screenshot 2024-06-05 at 10 27 16 PM

**Comment by ahmedGaber93** [comment]
@b4s36t4 I don't think this related to your PR.

**Comment by b4s36t4** [comment]
Thanks for checking @ahmedGaber93 .

**Comment by b4s36t4** [comment]
@ahmedGaber93 Please go ahead with your review and testing. I have uploaded the videos as well (except android chrome, setting up ssl certificated would be done soon). 

> Commit verification is also being handled, just wanted to unblock you.

**Comment by ahmedGaber93** [comment]
@b4s36t4 I think you missed those two places
https://github.com/Expensify/App/blob/afe64e55b34acad404c7cd0af945f8142e8380e3/src/libs/ReportUtils.ts#L2897

https://github.com/Expensify/App/blob/afe64e55b34acad404c7cd0af945f8142e8380e3/src/libs/ReportUtils.ts#L2919

**Comment by b4s36t4** [comment]
because `keyof CurrencyList` is of type `string`. which itself again leaves it to string.

Although typescript doesn't catch this I had a validation condition inside the function which will catch there even if we miss here.

**Comment by b4s36t4** [comment]
@ahmedGaber93 please check now updated, sorry for missing it out!.

**Comment by ahmedGaber93** [comment]
https://github.com/Expensify/App/blob/fcfef92de701acca9f4a1d18a8d2e64a356048bb/src/types/onyx/Currency.ts#L24

Yes, `CurrencyList` key is string as the list seem come from backend and save to onyx not static in the code

File: src/libs/CurrencyUtils.ts [hunk 1]
```diff
@@ -110,11 +111,20 @@ function convertToFrontendAmountAsString(amountAsInt: number | null | undefined)
  * @param amountInCents – should be an integer. Anything after a decimal place will be dropped.
  * @param currency - IOU currency
  */
-function convertToDisplayString(amountInCents = 0, currency: string = CONST.CURRENCY.USD): string {
+function convertToDisplayString(amountInCents = 0, currency: keyof CurrencyList | undefined = CONST.CURRENCY.USD): string {
```
Comment (ahmedGaber93) [inline src/libs/CurrencyUtils.ts:114]: https://github.com/Expensify/App/blob/fcfef92de701acca9f4a1d18a8d2e64a356048bb/src/types/onyx/Currency.ts#L24

It seems `keyof CurrencyList ` not work as expected as the list come from backend and save to onyx not static in the code so the `keyof CurrencyList` is `string[]`

I think we need to revert this change.

File: src/libs/CurrencyUtils.ts [hunk 2]
```diff
@@ -110,11 +111,20 @@ function convertToFrontendAmountAsString(amountAsInt: number | null | undefined)
...
     const convertedAmount = convertToFrontendAmountAsInteger(amountInCents);
+    /**
+     * Handle currency empty which can break the application.
+     * We're doing validation here so that we never send the value which will break the application.
+     * More: https://github.com/Expensify/App/issues/43004
+     */
+    let validatedCurrency = currency;
+    if (!currency || currency.length === 0 || currency.length > 3) {
+        validatedCurrency = CONST.CURRENCY.USD;
+    }
```
Comment (ahmedGaber93) [inline src/libs/CurrencyUtils.ts:124]: I think we can simplify this condition to `!currency` and this will check for `undefined` or `''`
```javascript
let currencyWithFallback = currency;
if (!currency) {
    currencyWithFallback = CONST.CURRENCY.USD;
}
```
Also, I think it is self-explanatory

Comment (b4s36t4) [reply src/libs/CurrencyUtils.ts:124]: Yea, but my idea is to make it explain the value shouldn't be empty string as well.

File: src/libs/CurrencyUtils.ts (hunk 1 above)
Comment (b4s36t4) [reply src/libs/CurrencyUtils.ts:114]: Screenshot 2024-06-06 at 2 30 27 AM

I don't think so, we can keep it, I don't think there's no issue with this.

**Comment by ahmedGaber93** [comment]
@b4s36t4 I think we need to simplify Tests to


1. Open any report
2. Click (+) icon > Submit expense then submit an expense
3. Open the expense report that you created it in step 2 and Verify the app is not crashed
- extra steps for Web and Desktop only
4. Open your browser console and type `Onyx.merge("report_<your_report_id>",{currency: null})` then press Enter. You can get `your_report_id` from browser link.
5. Verify the app is not crashed

**Comment by b4s36t4** [comment]
Agree, updating.!

File: src/libs/CurrencyUtils.ts (hunk 2 above)
Comment (ahmedGaber93) [reply src/libs/CurrencyUtils.ts:124]: > Yea, but my idea is to make it explain the value shouldn't be empty string as well.

Ok, we can simplify the comment, maybe like this `// Fallback currency to USD if it empty string or undefined`

File: src/libs/CurrencyUtils.ts (hunk 1 above)
Comment (ahmedGaber93) [reply src/libs/CurrencyUtils.ts:114]: Yes, there are no issues with it, But it not works as expect. If you pass currency as `''` it not display type error, so we need to revert it to `string`

File: src/libs/CurrencyUtils.ts (hunk 2 above)
Comment (b4s36t4) [reply src/libs/CurrencyUtils.ts:124]: Done, updated.!

Comment (ahmedGaber93) [reply src/libs/CurrencyUtils.ts]: @b4s36t4 I think this syntax is better
```javascript
let currencyWithFallback = currency;
if (!currency) {
    currencyWithFallback = CONST.CURRENCY.USD;
}
```

**Comment by ahmedGaber93** [comment]
@b4s36t4 We still need to update https://github.com/Expensify/App/pull/43133#discussion_r1628471064 and https://github.com/Expensify/App/pull/43133#discussion_r1628466928.
And please remove this part as it already cover by step 4 and maybe misleads QA team as it should be `null` not `INR`.

Screenshot 2024-06-06 at 12 42 02 AM

Thanks for your patience.

**Comment by ahmedGaber93** [comment]
I will complete the checklist in the morning.

**Comment by ahmedGaber93** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by ahmedGaber93** [comment]
@b4s36t4 this issue is urgent, are you able to update this https://github.com/Expensify/App/pull/43133#issuecomment-2151049696 today?

**Comment by b4s36t4** [comment]
@ahmedGaber93 Sorry I went to sleep at that time. I have updated the changes you requested can you please check again?

File: src/libs/CurrencyUtils.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@ import Onyx from 'react-native-onyx';
 import CONST from '@src/CONST';
 import type {OnyxValues} from '@src/ONYXKEYS';
 import ONYXKEYS from '@src/ONYXKEYS';
+import type {CurrencyList} from '@src/types/onyx';
```
[the remark this answers is not in the document]
Comment (b4s36t4) [reply src/libs/CurrencyUtils.ts]: Sorry, missed it.!

**Review by ahmedGaber93 — APPROVED** [review]
LGTM!

**Review by srikarparsi — APPROVED** [review]
code looks good and tests well, thanks for the quick changes

**Comment by srikarparsi** [comment]
QA, if you're not able to follow the test steps (deals with the browser console), please tag me and I can help you QA it.
